### PR TITLE
#159 Added action for Johnzon upgrade for Spring boot migration 2.7 t…

### PR DESCRIPTION
**Pull request by ravig-kant**
…o 3.x

### CODE REVIEW

**Review by fabapp2 — CHANGES_REQUESTED** [review]
Hi @ravig-kant 
thanks a lot for you contribution, very much appreciated!
I had a few small comments, could you have a look?
Looking forward to merge your PR.

**Comment by BoykoAlex** [comment]
Shouldn't the recipe go in **spring-boot-upgrade-30** project under the root? Or at lease some Rewrite dependent portion of the SBM recipe?

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade_27_30/conditions/JohnzonDependencyCondition.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+public class JohnzonDependencyCondition implements Condition {
+
+    @Override
+    public String getDescription() {
+        return "Checks if any of the class in the project imports Apache Jhonzon classes";
+    }
+
+    @Override
+    public boolean evaluate(ProjectContext context) {
+        return context.getProjectJavaSources()
```
[the remark this answers is not in the document]
Comment (ravig-kant) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade_27_30/conditions/JohnzonDependencyCondition.java:19]: I think I have to consider both conditions. It's possible that dependency is spring managed and hence doesn't reflect in pom. However, the project is nonetheless using Johnzon classes.

File: components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependenciesTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.sbm.build.api.Dependency;
+import org.springframework.sbm.engine.context.ProjectContext;
+import org.springframework.sbm.project.resource.TestProjectContext;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class Boot_27_30_UpgradeReplaceJohnzonDependenciesTest {
+
+    private static final String JHONZON_UNSUPPORTED_DEPENDENCY = "org.apache.johnzon:johnzon-core:1.2.11";
```
[the remark this answers is not in the document]
Comment (ravig-kant) [reply components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependenciesTest.java]: Thanks. Code updated.

File: components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependencies.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+public class Boot_27_30_UpgradeReplaceJohnzonDependencies extends AbstractAction {
+
+    private static final String JHONZON_DEPENDENCY_PATTERN = "org.apache.johnzon:johnzon-core";
+    private static final String JHONZON_DEPENDENCY = "org.apache.johnzon:johnzon-core:1.2.18-jakarta";
+
+    @Override
+    public void apply(ProjectContext context) {
+        BuildFile buildFile = context.getBuildFile();
+        buildFile.removeDependenciesMatchingRegex(JHONZON_DEPENDENCY_PATTERN);
```
[the remark this answers is not in the document]
Comment (ravig-kant) [reply components/sbm-recipes-boot-upgrade/src/main/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependencies.java]: Thanks. Code updated.

File: components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependenciesTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                                                                    + "}";
+
+
+    @Test
+    public void givenProjectWithSelfManagedJohnzonDependency_migrate_expectJohnzonDependencyUpdated(){
+        ProjectContext projectContext = getProjectContextWithJohnzonDependency(JAVA_CLASS_WITH_JHONZON_IMPORT);
+        Boot_27_30_UpgradeReplaceJohnzonDependencies upgradeReplaceJohnzonDependencies = new Boot_27_30_UpgradeReplaceJohnzonDependencies();
+        upgradeReplaceJohnzonDependencies.apply(projectContext);
+
+        assertTrue(projectContext
```
[the remark this answers is not in the document]
Comment (ravig-kant) [reply components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependenciesTest.java]: Thanks. Code updated.

File: components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependenciesTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                .withBuildFileHavingDependencies(JHONZON_UNSUPPORTED_DEPENDENCY)
+                .withJavaSources(javaSources)
+                .build();
+    }
+
+    private ProjectContext getProjectContextWithoutJohnzonDependency(String... javaSources){
+        return TestProjectContext.buildProjectContext()
+                .withDummyRootBuildFile()
+                .withJavaSources(javaSources)
+                .addRegistrar(new SpringBootApplicationPropertiesRegistrar(new SpringApplicationPropertiesPathMatcher()))
```
[the remark this answers is not in the document]
Comment (ravig-kant) [reply components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependenciesTest.java]: Thanks. Code updated.

**Comment by ravig-kant** [comment]
Hi @fabapp2 

I have a question about handling Johnzon dependencies for multi-module projects. In such cases should we not be updating the parent projects dependency management section? Also for spring-managed cases, we will have to add a dependency to the child module.

> Hi @ravig-kant thanks a lot for you contribution, very much appreciated! I had a few small comments, could you have a look? Looking forward to merge your PR.

Thanks
Ravi

**Comment by fabapp2** [comment]
> Shouldn't the recipe go in **spring-boot-upgrade-30** project under the root? Or at lease some Rewrite dependent portion of the SBM recipe?

I'd say it's the right module for now. @sanagaraj-pivotal killed the `spring-boot-upgrade-30´ module as this is basically `sbm-openrewrite` where the resources were moved to. As the migration is done in plain SBM for now this module is ok imho.

**Comment by fabapp2** [comment]
> Hi @fabapp2
> 
> I have a question about handling Johnzon dependencies for multi-module projects. In such cases should we not be updating the parent projects dependency management section? Also for spring-managed cases, we will have to add a dependency to the child module.
> 
> > Hi @ravig-kant thanks a lot for you contribution, very much appreciated! I had a few small comments, could you have a look? Looking forward to merge your PR.
> 
> Thanks Ravi

Hi @ravig-kant 

I am currently working on multi-module support.
Moving fixed set versions into a `dependencyManagement` section is something I could see as a separate recipe.
For this issue it should be enough to just set the version and classifier for all johnson dependencies found.

Making this multi-module aware could be achieved like so:
```
context.getApplicationModules().stream()
                .map(ApplicationModule::getBuildFile)
                .filter(b -> b.hasDeclaredDependencyMatchingRegex("..."))
                .forEach(this::doSomethingWithBuildFileContainingJohnzonDependency);
```

**Comment by fabapp2** [comment]
Hi @ravig-kant,
only a few tweaks are required to merge your PR to `main`:

- [ ] Remove the checks for imports, just having the dependency without a version (managed by Boot) should trigger the replacement with 
```
<dependency>
            <groupId>org.apache.johnzon</groupId>
            <artifactId>johnzon-core</artifactId>
            <version>1.2.18</version>
	    <classifier>jakarta</classifier>
        </dependency>
```

- [ ] Add support for classifier in `Dependency.fromCoordinates()`
- [ ] Making the Action multi module aware by checking all `ApplicationModule`s
- [ ] Adjust the test to expect `jakarta` as classifier in the added dependency

**Review by fabapp2 — APPROVED** [review]

**Comment by fabapp2** [comment]
Thanks, @ravig-kant for your contribution 🎉

File: components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependenciesTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            "    <modelVersion>4.0.0</modelVersion>\n" +
+            "    <groupId>com.example</groupId>\n" +
+            "    <artifactId>dummy-root</artifactId>\n" +
+            "    <version>0.1.0-SNAPSHOT</version>\n" +
+            "    <packaging>jar</packaging>\n" +
+            "    <dependencies>\n" +
+            "        <dependency>\n" +
+            "            <groupId>org.apache.johnzon</groupId>\n" +
+            "            <artifactId>johnzon-core</artifactId>\n" +
+            "            <version>1.2.18-jakarta</version>\n" +
```
[the remark this answers is not in the document]
Comment (ravig-kant) [reply components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/actions/Boot_27_30_UpgradeReplaceJohnzonDependenciesTest.java]: I think your comments and my next upload crossed. I added the test for the new case. Thanks for the review. Hopefully, I will improve and churn out better quality next time. I will pick up #225 now.

**Pull request merged**